abstractions/typescript/src/authentication/allowedHostsValidator.ts

```diff
@@ -0,0 +1,60 @@
+/** Maintains a list of valid hosts and allows authentication providers to check whether a host is valid before authenticating a request */
+export class AllowedHostsValidator {
+    private allowedHosts: Set<string>;
+    /**
+     * Creates a new AllowedHostsValidator object with provided values.
+     * @param allowedHosts A list of valid hosts.  If the list is empty, all hosts are valid.
+     */
+    public constructor(...allowedHosts: string[]) {
+        this.allowedHosts = new Set(allowedHosts?.map(x => x?.toLowerCase()));
+    }
+    /**
+     * Gets the list of valid hosts.  If the list is empty, all hosts are valid.
+     * @returns A list of valid hosts.  If the list is empty, all hosts are valid.
+     */
+    public getAllowedHosts(): string[] {
+        return Array.from(this.allowedHosts);
+    }
+    /**
+     * Sets the list of valid hosts.  If the list is empty, all hosts are valid.
+     * @param allowedHosts A list of valid hosts.  If the list is empty, all hosts are valid.
+     */
+    public setAllowedHosts(...allowedHosts: string[]): void {
+        this.allowedHosts = new Set(allowedHosts?.map(x => x?.toLowerCase()));
+    }
+    /**
+     * Checks whether the provided host is valid.
+     * @param url The url to check.
+     */
+    public isUrlHostValid(url: string): boolean {
```

Comment:
Why use a class instead of simply exporting a util function?

Reply:
because we need something to hold the list of valid hostnames?

Reply:
Hmm.. I think that would be necessary for Java or dot net. 

Currently we maintain this as util functions: 
 
https://github.com/microsoftgraph/msgraph-sdk-javascript/blob/3eab6727853b1977bd8046fa3bab8e95dafe6f40/src/GraphRequestUtil.ts#L68

https://github.com/microsoftgraph/msgraph-sdk-javascript/blob/9441382f7963fa782cd30a103af2632d7585e0a0/src/Constants.ts#L28


Related to the comment below we could validate the url using a single util function.

Reply:
the main difference being the list of hosts is maintained on a global context object, which is a mix of everything (request options, hosts, etc...) vs a self contained (encapsulated) object.
https://github.com/microsoftgraph/msgraph-sdk-javascript/blob/3eab6727853b1977bd8046fa3bab8e95dafe6f40/src/IContext.ts#L29

I do have a preference for encapsulation over a mix "context" object. What do you think?

Reply:
I am with you about not storing hosts in context object. 

When I thinking of the flow for Graph JS SDK we would go 

1. ```
Client.init({
authProvider= new AzureAuthenticationProvider(credential, user_custom_hosts);
});
```
2. The user_custom_hosts would not contain the Graph hosts.  

3. At the time of initialization I would append the Graph hosts to this user_custom_hosts. 

Is the use case valid? 
If that is the case, the allowed hosts should be public property so that it can be modified. 


We can continue this discussion in #1063. This conversation need not block this PR.

Reply:
right, we're probably missing a getter to the allowed hosts validator in the access token provider interface for the chain to be complete to people can access things afterwards.

Reply:
added the getter

Reply:
CC @Ndiritu for PHP, sorry I already had merged your PR

Reply:
also CC @samwelkanda for the ongoing work in #925

Reply:
> right, we're probably missing a getter to the allowed hosts validator in the access token provider interface for the chain to be complete to people can access things afterwards.


How about have a public function `AppendHosts` in the hosts validator class.

Reply:
if we add an append host (which could be done by Get, append to the result, set) we'd also need to add a remove host method for consistency. All of that can already be done with the current API surface, I don't think it'll add much value to the current offering.